crates/rb-allocator/src/lib.rs

```diff
@@ -0,0 +1,34 @@
+//! Allocator which reports mem usage to the Ruby GC
+//!
+//! This crate exports, one type, `RbAllocator`, which implements the `GlobalAlloc` trait. It is a
+//! simple wrapper over the system allocator which reports memory usage to Ruby using
+//! `rb_gc_adjust_memory_usage`
+
+use rb_sys::{rb_gc_adjust_memory_usage, ssize_t};
+use std::alloc::{GlobalAlloc, Layout, System};
+
+pub struct RbAllocator;
+
+unsafe impl GlobalAlloc for RbAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let ret = System.alloc(layout);
+        if !ret.is_null() {
```

Comment:
If it does return null, then ruby's allocation functions (e.g. ruby_xmalloc) will invoke the garbage collector and retry.  I would have expected something like `rb_gc()` to be called to handle this condition.

```suggestion
        if ret.is_null() {
          rb_gc();
          ret = System.alloc(layout);
        }
        if !ret.is_null() {
```

along with an import of `rb_gc` and `ret` being made mutable.